### ISSUE TITLE
1.3.x samplerate bug "41000"

### ISSUE THREAD

**Issue by thumpclang** [danomatika/PdParty#100]
iPad 5th generation running iPadOS 16.1.1

Starting in PdParty 1.3.0 I noticed that playback of an array via tabread4~ in a subpatch is playing back at a higher pitch, however my samplerate is set to 44.1k as it should be. My patch works as expected on my Mac in Pd-vanilla, and I believe was working correctly in PdParty pre-1.3.0 (although I don't know if I can downgrade to double check this).

Attaching a simple test patch; the first screenshot is on Mac, the second on iPad in PdParty showing the unexpected frequency result:

[record_playback_test.pd.zip](https://github.com/danomatika/PdParty/files/10317722/record_playback_test.pd.zip)

<img width="829" alt="pdvanillascreenshot" src="https://user-images.githubusercontent.com/88018389/209907845-4f1ab93c-9fbf-4fad-bd0d-362e20a6d614.png">

![pdpartyscreenshot](https://user-images.githubusercontent.com/88018389/209907858-a756c367-28cf-4c95-95ea-63774feaf4d2.PNG)

Also, interestingly, TriSamp records & plays back at the expected pitch/samplerate; I'm not sure what the difference would be between my test patch & the recording/playback in TriSamp, but perhaps this may help narrow down the possible bug.

**Comment by danomatika** [comment]
I updated Trisamp to use [samplerate~] since the samplerate is now settable.

**Comment by danomatika** [comment]
Using the output off [samplerate~] for tabread playback gives me the correct playback frequency. I can't account for why this would not work at the specific samplerate, especially when audio unit is running at 44.1k and confirms the output stream is also 44.1k, ie. audio rate conversion is being handled by the system.

I'm surprised it might have worked for you before since PdParty 1.2 was not enforcing samplerate behavior but had to use whatever the system provided, ie. 48k.

@Spacechild1 @millerpuckette Any reason why sigmund would be acting strangely in this case?

**Comment by danomatika** [comment]
Here is an update that works for me, although it doesn't answer the question as to the original behavior:

<img width="349" alt="Screen Shot 2022-12-29 at 11 50 41 AM" src="https://user-images.githubusercontent.com/480637/209940953-76937dcb-1481-4d41-9905-aed62757270e.png">

<img width="672" alt="Screen Shot 2022-12-29 at 11 50 28 AM" src="https://user-images.githubusercontent.com/480637/209940964-18a11417-5602-4fdf-ba1d-fb094803c2cb.png">

[record_playback_test_sr.pd.zip](https://github.com/danomatika/PdParty/files/10318995/record_playback_test_sr.pd.zip)

**Comment by danomatika** [comment]
>(although I don't know if I can downgrade to double check this)

@thumpclang Downgrading is not possible, unfortunately.

**Comment by thumpclang** [comment]
Aha! I tried your modified patch:
![sampleratescreenshot](https://user-images.githubusercontent.com/88018389/209999560-1a71a412-1dcf-4164-adb6-27cf2ef1e9f8.PNG)

I [searched](https://github.com/danomatika/PdParty/search?q=41000) for "41000" in PdParty's Github repo, is this the typo we're looking for?:

src/SettingsViewController.m
41   sampleRateValues = @[@(48000), @(41000), @(96000)];

In any case, I certainly should get in the habit of using a samplerate-agnostic method for playback, I'm sure it'll come back to bite me again someday. (and, also seems like sigmund isn't doing anything unexpected here)

**Comment by danomatika** [comment]
> I [searched](https://github.com/danomatika/PdParty/search?q=41000) for "41000" in PdParty's Github repo, is this the typo we're looking for?:
> 
> src/SettingsViewController.m 41 sampleRateValues = @[@(48000), @(41000), @(96000)];

No, the order is correct. I modeled it after Pd vanilla's audio dialog which lists 48k first. This array is only used for the UI radio button, ie. by index. The value itself is saved in the settings and used when setting samplerate on start.

**Comment by thumpclang** [comment]
The order makes sense, but 41000 < 44100, yes? :)

**Comment by danomatika** [comment]
> The order makes sense, but 41000 < 44100, yes? :)

I don't understand the question.

**Comment by thumpclang** [comment]
Apologies, I meant that 41000 is not a standard samplerate, but that seems to be what PdParty is setting the samplerate to when selecting the "44.1k" button, as confirmed by my screenshot. I was just guessing that perhaps that line of code in src/SettingsViewController.m should include "@(44100)" in place of "@(41000)", does that seem right? Or is 41000 actually the intended samplerate?

**Comment by danomatika** [comment]
Ok, I see what you mean. Yes this is a typo and I obviously didn't notice the difference in your previous post. I am wrapping up a 1.3.2 which fixes #99, so I can add this fix as well.

**Comment by danomatika** [comment]
I can confirm running with "41000" causes the issue. Fixing the value and reselecting 44.1k in the app gives me the correct values from sigmund in the original test patch. This was a "copy-pasta" error when creating the array. Most people would not have noticed this since 48k is the default and if the value loaded from settings isn't in the array, the default is used.

**Comment by danomatika** [comment]
The fix for this is 5e588efd9a74c323d51a0b56e92131f380245981 which is now in PdParty 1.3.2.

**Comment by thumpclang** [comment]
Great, thanks for the fix, and also for shaking out the clue that led to finding the problem! I was utterly confused before considering that the bug could be somewhere other than my patch, so this is a big relief.

And indeed, there's something about the number 44100 that I find very typo & error prone; can't count how many times I've mistyped or miscalculated when converting for various uses. Long story short, I learned to program in Pd on a 2008 MacBook, at which time I never imagined needing a higher samplerate, and have simply kicked the can down the road ever since... I'm going to take this experience as an impetus to finally go through all of my old patches with 44.1k-dependent calculations and finally update everything to use [samplerate~] as you're wisely doing.
